x/provider/handler/handler_test.go

```diff
@@ -0,0 +1,129 @@
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+	dbm "github.com/tendermint/tm-db"
+
+	"github.com/ovrclk/akash/app"
+	"github.com/ovrclk/akash/testutil"
+	"github.com/ovrclk/akash/x/provider/handler"
+
+	"github.com/ovrclk/akash/x/provider/keeper"
+	"github.com/ovrclk/akash/x/provider/types"
+)
+
+type testuite struct {
+	ms      sdk.CommitMultiStore
+	ctx     sdk.Context
+	keeper  keeper.Keeper
+	handler sdk.Handler
+}
+
+func setupTestContext(t *testing.T) *testuite {
+	suite := &testuite{}
+	testutil.AccAddress(t)
+	keyProvider := sdk.NewTransientStoreKey(types.StoreKey)
+
+	db := dbm.NewMemDB()
+	suite.ms = store.NewCommitMultiStore(db)
+	suite.ms.MountStoreWithDB(keyProvider, sdk.StoreTypeIAVL, db)
+
+	err := suite.ms.LoadLatestVersion()
+	require.NoError(t, err)
+
+	suite.ctx = sdk.NewContext(suite.ms, abci.Header{}, true, log.NewNopLogger())
+
+	suite.keeper = keeper.NewKeeper(app.MakeCodec(), keyProvider)
+	suite.handler = handler.NewHandler(suite.keeper)
+
+	return suite
+}
+
+func TestProviderCreate(t *testing.T) {
+	suite := setupTestContext(t)
+
+	msg := types.MsgCreateProvider{
+		Owner:   testutil.AccAddress(t),
+		HostURI: "host",
+	}
+
+	_, err := suite.handler(suite.ctx, msg)
+	require.NoError(t, err)
+
+	_, err = suite.handler(suite.ctx, msg)
+	require.Error(t, err)
+	require.True(t, errors.Is(err, types.ErrProviderExists))
+}
+
+func TestProviderUpdateExisting(t *testing.T) {
+	suite := setupTestContext(t)
+
+	addr := testutil.AccAddress(t)
+
+	createMsg := types.MsgCreateProvider{
+		Owner:   addr,
+		HostURI: "host",
+	}
+
+	updateMsg := types.MsgUpdateProvider{
+		Owner:   addr,
+		HostURI: "host",
+	}
+
+	err := suite.keeper.Create(suite.ctx, types.Provider(createMsg))
+	require.NoError(t, err)
+
+	_, err = suite.handler(suite.ctx, updateMsg)
+	require.NoError(t, err)
+}
+
+func TestProviderUpdateNotExisting(t *testing.T) {
+	suite := setupTestContext(t)
+	msg := types.MsgUpdateProvider{
+		Owner:   testutil.AccAddress(t),
+		HostURI: "host",
+	}
+
+	_, err := suite.handler(suite.ctx, msg)
+	require.Error(t, err)
+	require.True(t, errors.Is(err, types.ErrProviderNotFound))
+}
+
+func TestProviderDeleteExisting(t *testing.T) {
+	suite := setupTestContext(t)
+
+	addr := testutil.AccAddress(t)
+
+	createMsg := types.MsgCreateProvider{
+		Owner:   addr,
+		HostURI: "host",
+	}
+
+	deleteMsg := types.MsgDeleteProvider{
+		Owner: addr,
+	}
+
+	err := suite.keeper.Create(suite.ctx, types.Provider(createMsg))
+	require.NoError(t, err)
+
+	_, err = suite.handler(suite.ctx, deleteMsg)
+	require.Error(t, err)
+}
+
+func TestProviderDeleteNonExisting(t *testing.T) {
+	suite := setupTestContext(t)
+	msg := types.MsgDeleteProvider{
+		Owner: testutil.AccAddress(t),
+	}
+
+	_, err := suite.handler(suite.ctx, msg)
+	require.Error(t, err)
+	require.True(t, errors.Is(err, types.ErrProviderNotFound))
```

Comment:
does `require` or `assert` have an `ErrorIs()` or similar?

Reply:
neither has. there was suggestion to make `assert.Wraps` to the `testify` package but still as proposal